Try/e2e ci

bsessions85 opened the pull request:
I added the running of e2e tests to the travis build config. 

**How it works:**
The run-e2e.sh script does some setup and then creates a new wordpress instance on serverpilot.  The script logs in to the new instance and runs sp-config script that updates the db and installs WC from the branch.
After that there is a site to test against so we run the tests.
When the tests are complete the site is deleted from serverpilot.

**How it is set up**
It is currently set to run on one job in the build matrix. We could run on all, but I was getting conflicts with different build agents not having the correct dependencies . 

The variables are set as secure environment variables in Travis. 

**Notes**
I had to move the e2e tests in to a subdirectory so that mocha wouldn't keep trying to run the stuff in the scripts directory. 

We'll definitely want to squash this one on merge. It took a lot of testing on Travis to get everything working!

Commit:
Attempting to get tests to run in Travis CI

Commit:
Changing to ES5 and setting config dir

Commit:
Configuration changes

Commit:
Update config path and add serverpilot package

Commit:
Set plugin by branch

Commit:
Debugging SHA

Commit:
Changing to build id

Commit:
Setting up ssh to sp

Commit:
Fixing broken script

Commit:
debugging script

Commit:
adjusting paths

Commit:
Trying to get bash output

Commit:
Fixing typo in path

Commit:
Fixing broken app

Commit:
Reorganizing

Commit:
Changing config

Commit:
Debugging the site deleting

Commit:
back to ES6

Commit:
Param numbers off

Commit:
Site is still being deleted

Commit:
More troubleshooting

Commit:
Try to to plugin install better

Commit:
Plugin not activating properly

Commit:
Reducing runs

Commit:
Still deleting on its own

Commit:
require chrome

Commit:
Trying grunt task instead

Commit:
Exiting the ssh session

Commit:
Changing structure to keep mocha from running config scripts

Commit:
Run tests with npm

Commit:
Fixing npm command

Commit:
debugging

Commit:
Setting base url variable

Commit:
Trying variables in config

Commit:
Fixing config with variables

Commit:
missed a comma in json

Commit:
key type wrong

Commit:
Fixing config

Commit:
Fixing config

Commit:
trying single test

Commit:
Plugin not staying activated

Commit:
adding line for xvfb

Commit:
breaking out script for so tests fail builed

Commit:
fixed

Commit:
fxing again

Commit:
trying fix

Commit:
turning off debug

Commit:
trying with grunt

Commit:
Attempting new DB approach

Commit:
attempting to fix issue

Commit:
db wasn't created

Commit:
Troubleshooting

Commit:
forcing

Commit:
escaping characters

Commit:
setting memory

Commit:
changing install order

Commit:
trying more changes

Commit:
trying to fail build if tests fail

Commit:
another attempt

Commit:
more attempt

Commit:
reverting

Commit:
trying script re-org

Commit:
changing permissions

Commit:
trying this

Commit:
broke it

Commit:
space instead of tab

Commit:
another try

Commit:
running just one test

Commit:
debug

Commit:
trying more

Commit:
more

Commit:
I think I have it

Commit:
trying another approach

Commit:
and again

Commit:
Trying some more configurations

Commit:
More testing

Commit:
Prepare for PR

Commit:
Merge branch 'master' into try/e2e-CI

# Conflicts:
#	tests/e2e-tests/tests/single-product.js

Commit:
post merge testing

Commit:
Fixing broken refactor

Commit:
Removing test line

Commit:
Renaming script

claudiulodro commented:
Seems to work great! Thanks!

rodrigoprimo committed:
Remove unnecessary spaces

rodrigoprimo commented on .travis.yml:
```diff
@@ -35,9 +38,12 @@ before_script:
 
 after_script:
   - bash tests/bin/travis.sh after
+  - export NODE_CONFIG_DIR="./tests/e2e-tests/config"
```
What do you think if we move this line and the line below to `tests/bin/travis.sh`? It seems to me that it would be better to keep all the code that is executed "after_script" in a single place.

rodrigoprimo commented on .travis.yml:
```diff
@@ -35,9 +38,12 @@ before_script:
 after_script:
   - bash tests/bin/travis.sh after
+  - export NODE_CONFIG_DIR="./tests/e2e-tests/config"
+  - ./tests/e2e-tests/scripts/wp-serverpilot-delete.js
```
Should we maybe add a check to call this script only when `RUN_E2E` is true? This way we would avoid calling it on jobs where we don't run e2e tests. This is producing the error below:

https://travis-ci.org/woocommerce/woocommerce/jobs/326418964#L303

rodrigoprimo commented on tests/e2e-tests/scripts/run-e2e.sh:
```diff
@@ -0,0 +1,29 @@
+	sh -e /etc/init.d/xvfb start
+ 	sleep 3
+	npm install
+	export NODE_CONFIG_DIR="./tests/e2e-tests/config"
```
This is probably a Node beginner question and not directly related to your PR, but, in case you know the answer, I'm curious to know why we need to declare `NODE_CONFIG_DIR` everywhere and I wonder if there is a way to do it only once and avoid repetition?

rodrigoprimo commented on package.json:
```diff
@@ -46,6 +46,7 @@
     "grunt-wp-i18n": "~1.0.1",
     "istanbul": "^1.0.0-alpha",
     "mocha": "^3.0.2",
+    "serverpilot": "0.1.2",
```
Could you share a little bit of why you decided to use serverpilot to create a WC instance to run the tests? I'm just curious to know a bit about your decision making process. I was under the impression that we would be able to configure a WC instance using just Travis but I'm probably wrong :) Or maybe it is possible but serverpilot is easier? Thanks.

rodrigoprimo commented on tests/e2e-tests/scripts/run-e2e.sh:
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
```
Could we maybe add one or two lines documenting what is this script purpose? I think it would be useful to do the same to `sp-config.sh`, `wp-serverpilot-delete.js` and `wp-serverpilot-init.js`

Also, I'm unsure if this script should be in this directory with the rest of the e2e related scripts or in `tests/bin` where we keep all the other scripts that are called by Travis. Thoughts?

rodrigoprimo reviewed (commented):
Super excited to see this PR. I left a few comments with some questions. Thanks for your work.

bsessions85 commented:
@rodrigoprimo  You make some good points. I will make some updates and look in to running on the travis container itself. I originally went with serverpilot because that is what we ended up using for jetpack in CircleCI. Travis does seem to have better options for running locally, so I will give that a try. Thanks!

Commit:
Trying local install

Commit:
Merge branch 'try/e2e-CI' of github.com:woocommerce/woocommerce into try/e2e-CI

Commit:
Installing wp cli

Commit:
Try without installing

Commit:
locale

Commit:
checking localhost

Commit:
adding confgi first

Commit:
removing curl

Commit:
fixing data path

Commit:
fixing branch

Commit:
trying wp werver

Commit:
starting server

Commit:
changing port

Commit:
trying rackup in yml

Commit:
trying php server

Commit:
trying nginx

Commit:
trying sudo

Commit:
copy config

Commit:
trying to move wordpress install

Commit:
typo

Commit:
trying memory limit change

Commit:
troubleshooting

Commit:
getting closer

Commit:
another try

Commit:
more trying

Commit:
still 403ing

Commit:
permissions issue

Commit:
trying again

Commit:
troubleshoot

Commit:
changing permissions

Commit:
more permission changes

Commit:
trying to restart nginx

Commit:
trying conf changes

Commit:
troubleshooting nginx

Commit:
trying some more troubleshooting

Commit:
again

Commit:
Trying new setup for nginx

Commit:
Bad path

Commit:
missed script

Commit:
bad path

Commit:
trying different port

Commit:
missed a replacement

Commit:
troubleshooting wordpress install

Commit:
trying nginx stop first

Commit:
wrong command

Commit:
trying to see nginx config

Commit:
fixixng bad ports

Commit:
trying some changes

Commit:
removing variables

Commit:
change task order

Commit:
bad path to config

Commit:
changed config order

Commit:
another bad path

Commit:
wp install not working

Commit:
trying to skip email

Commit:
only other param I can think of

Commit:
troubleshooting

Commit:
change site url

Commit:
tests not working

Commit:
cleaning up

Commit:
Merge branch 'master' into try/e2e-CI

Commit:
Trying some configuration changes

Commit:
More Cleanup

Commit:
refactoring

Commit:
Missed a variable

Commit:
removing more unneeded resources

Commit:
more cleanup

Commit:
reverting package change

Commit:
cleaning

Commit:
got overzealous

Commit:
Merge branch 'master' into try/e2e-CI

Commit:
Update page objects

bsessions85 commented:
@rodrigoprimo I changed the format of this. I changed it to just run a wordpress instance in the Travis container so we aren't dependent on ServerPilot. I moved the tests back to where they were before and added a few config files to get nginx to run the instance. I added my setup tasks to the existing setup script and used a new script to start xvfb and run the tests.  It is much cleaner this way and all the tests are passing.

rodrigoprimo committed:
Use $TRAVIS_REPO_SLUG instead of hardcoded repository slug

Doing this to be able to run tests on forks

rodrigoprimo commented on tests/bin/install.sh:
```diff
@@ -123,6 +123,69 @@ install_db() {
 	mysqladmin create $DB_NAME --user="$DB_USER" --password="$DB_PASS"$EXTRA
 }
 
+install_e2e_site() {
```
Maybe for a future iteration, we can explore using the same WP install that is used for unit tests and just replace the database? This should be faster than installing another WP instance and I don't see any drawbacks, but I'm not sure if it is possible. `install_e2e_site()` is adding about 40 seconds to `install.sh` execution time (https://travis-ci.org/woocommerce/woocommerce/jobs/329686529#L518). When `install_e2e_site()` is not called, `install.sh` takes about 10 seconds to run.

I don't see this as a blocker, and I'm happy to merge this as is this week since next week we have the 3.3 release.

bsessions85 replied:
I like that idea. Would be good to look in to for future iterations

rodrigoprimo reviewed (approved):
LGTM. Thanks for working on this.

I made a minor change to be able to test e2e tests in CI on my WC fork (aea6e9751db6ef822d16185a782430d355d84b4a). I also left a comment with a suggestion that I think will improve the build execution time but we can work on that in the future on another PR.

On my fork, some e2e tests failed due to timeouts and we might need to improve them and in the meantime run e2e tests in the build job that is allowed to fail. But let's start as is and change if needed.
